### PR TITLE
Add OrbitalRotation operation

### PR DESCRIPTION
**Pull request by obliviateandsurrender**
**Context:**
This is the first part of the PRs aimed at adding the `GateFabric` template, which is based on the local, expressive, and quantum number preserving ansatz presented in [arXiv:2104.05695](https://arxiv.org/abs/2104.05695).

**Description of the Change:**
Function `OrbitalRotation` provides the implementation for the spin-adapted spatial orbital rotation gate which constitutes the gate required for developing the aforementioned template.
Relevant UnitTests for `OrbitalRotation` operation.

**Benefits:**
This operation will be required in the implementation of the `GateFabric` template. 

**Possible Drawbacks:**
No specific drawbacks.

**Related GitHub Issues:**
No related issues.

### CODE REVIEW

File: pennylane/ops/qubit/qchem_ops.py
Comment (ixfoduap) [inline pennylane/ops/qubit/qchem_ops.py:488]:
```suggestion
    For two neighbouring spatial orbitals :math:`\{|\Phi_{0}\rangle, |\Phi_{1}\rangle\}`, this operation
    performs the following transformation
```

File: pennylane/ops/qubit/qchem_ops.py
Comment (ixfoduap) [inline pennylane/ops/qubit/qchem_ops.py:490]: Really clever way to summarize what the operation is doing! :tada: :brain:

File: pennylane/ops/qubit/qchem_ops.py
Comment (ixfoduap) [inline pennylane/ops/qubit/qchem_ops.py:500]:
```suggestion
    * Gradient recipe: The ``OrbitalRotation`` operator satisfies the four-term parameter-shift rule
```

File: pennylane/ops/qubit/qchem_ops.py
Comment (ixfoduap) [inline pennylane/ops/qubit/qchem_ops.py:564]: Great to see `qml.math` in action!

File: tests/ops/test_qubit_ops.py
Comment (ixfoduap) [inline tests/ops/test_qubit_ops.py:3581]: How did you compute this state as to avoid using the `OrbitalRotation` operation? i.e., how do you know it's the correct state?

**Review by ixfoduap — CHANGES_REQUESTED** [review]
Looks great! Amazing job!

I didn't find any issues, so I'm happy to approve from my side once a small change in the docstrings is made

File: pennylane/ops/qubit/qchem_ops.py
Comment (obliviateandsurrender) [reply pennylane/ops/qubit/qchem_ops.py]: Updated in the docstring.

File: pennylane/ops/qubit/qchem_ops.py
Comment (obliviateandsurrender) [reply pennylane/ops/qubit/qchem_ops.py]: Updated in the docstring.

File: tests/ops/test_qubit_ops.py
Comment (obliviateandsurrender) [reply tests/ops/test_qubit_ops.py:3581]: I'd computed this state manually using the matrix definition of the gate.

File: pennylane/ops/qubit/qchem_ops.py
Comment (soranjh) [inline pennylane/ops/qubit/qchem_ops.py:513]: I think this should be:
`qml.BasisState(np.array([1, 1, 0, 0]), wires=[0, 1, 2, 3])`

File: tests/ops/test_qubit_ops.py
Comment (soranjh) [inline tests/ops/test_qubit_ops.py:3695]: Do we need to test different `diff_method`s here?

**Review by soranjh — APPROVED** [review]
The PR looks good to me @obliviateandsurrender, just left two comments. I am approving as I have already reviewed the new addition in a previous [PR](https://github.com/PennyLaneAI/pennylane/pull/1626).

File: pennylane/ops/qubit/qchem_ops.py
Comment (obliviateandsurrender) [reply pennylane/ops/qubit/qchem_ops.py]: Yeah, I have fixed it now.

File: tests/ops/test_qubit_ops.py
Comment (obliviateandsurrender) [reply tests/ops/test_qubit_ops.py]: Yes, we should. I have now added the `diff_method` for this unit test as well.

File: pennylane/ops/qubit/qchem_ops.py
Comment (josh146) [reply pennylane/ops/qubit/qchem_ops.py:488]: I agree!! really nice docstring 🙂 

One suggestion: it would be nice to briefly *describe* the image (in this case, I wasn't originally sure what `G(theta)` referred to, I assume givens rotations?)

Comment (josh146) [reply pennylane/ops/qubit/qchem_ops.py:488]: Actually, now that I look at it again: it might be clearer to remove the single-excitation image, and in the text below, simply say

```rest
:math:`G(\phi)` represents a single-excitation Given's rotation,
implemented in PennyLane as the :class:`~.SingleExcitation` operation.
```
This way a reader can easily click on the link to read more about the `SingleExcitation` operation.

File: pennylane/ops/qubit/qchem_ops.py
Comment (josh146) [inline pennylane/ops/qubit/qchem_ops.py:517]:
```suggestion
        >>> dev = qml.device('default.qubit', wires=4)
        >>> @qml.qnode(dev)
        ... def circuit(phi):
        ...     qml.BasisState(np.array([1, 1, 0, 0]), wires=[0, 1, 2, 3])
        ...     qml.OrbitalRotation(phi, wires=[0, 1, 2, 3])
        ...     return qml.state()
        >>> circuit(0.1)
```

File: pennylane/ops/qubit/qchem_ops.py
Comment (josh146) [reply pennylane/ops/qubit/qchem_ops.py:562]: Same! it is really nice not needing to modify `tf_ops.py` etc.

File: tests/ops/test_qubit_ops.py
Comment (josh146) [inline tests/ops/test_qubit_ops.py:3767]: @obliviateandsurrender could you add a PyTorch test here as well? Aside from that, great tests!

**Review by josh146 — CHANGES_REQUESTED** [review]
Really nice PR @obliviateandsurrender! Left a small comment regarding adding a pytorch gradient test. Apart from that, don't forget to add an entry to the changelog!

File: pennylane/ops/qubit/qchem_ops.py
Comment (obliviateandsurrender) [reply pennylane/ops/qubit/qchem_ops.py:476]: I agree. I have added this to the documentation now.

File: pennylane/ops/qubit/qchem_ops.py
Comment (obliviateandsurrender) [reply pennylane/ops/qubit/qchem_ops.py]: Fixed.

File: tests/ops/test_qubit_ops.py
Comment (josh146) [reply tests/ops/test_qubit_ops.py:3805]: Hi @obliviateandsurrender, nice catch -- pytorch can be very finicky with how matrices are defined.

No need to modify, but I'm curious if using `qml.math.stack` to first stack each row of the array, and then stack all rows, would have worked with PyTorch?

Comment (obliviateandsurrender) [reply tests/ops/test_qubit_ops.py:3805]: Hey @josh146, you're right. Using `qml.math.stack` indeed works! I think it would be much better to use this instead of the full decomposition.

Comment (josh146) [reply tests/ops/test_qubit_ops.py:3805]: Awesome! let me know when the master merge conflict is resolved, and I will give it another review :)

Comment (obliviateandsurrender) [reply tests/ops/test_qubit_ops.py:3805]: Hello, the merge conflict with the master is resolved now.

File: doc/releases/changelog-dev.md
Comment (josh146) [inline doc/releases/changelog-dev.md:103]: changelog entry looks 💯

File: pennylane/ops/qubit/qchem_ops.py
Comment (josh146) [reply pennylane/ops/qubit/qchem_ops.py:476]: Love it!

File: pennylane/ops/qubit/qchem_ops.py
Comment (josh146) [inline pennylane/ops/qubit/qchem_ops.py:579]:
```suggestion
        U = qml.math.stack([qml.math.stack(row) for row in matrix], axis=0)
```
I'm guessing this is the same thing?

**Review by josh146 — APPROVED** [review]
Good to go from my end @obliviateandsurrender, once @ixfoduap and @Jaybsoni also approve!

File: doc/releases/changelog-dev.md
Comment (ixfoduap) [inline doc/releases/changelog-dev.md:214]: Is this how you prefer to be referred to in a professional setting? Up to you, but I recommend adding your last name too

**Review by ixfoduap — APPROVED** [review]
:shipit:

**Pull request merged**
